### DISCUSSION TITLE
how to submit information to user mode instead of printing information in the kernel?

### DISCUSSION THREAD

**Discussion by caneatcansleep** [bpftrace/bpftrace#2354] [Q&A] [answered]
Before learning about bpftrace, I had been learning libbpf and BCC. I found that almost all [bpftrace programs](https://github.com/iovisor/bpftrace/tree/master/tools) use the `printf `function to print information. However, libbpf and BCC provide `perf buffer` to allow us to print information in user mode. 
My question is how to use `perf buffer` in bpftrace? Or is there a way to interact with the user mode in bpftrace, that is, not to print the information in the kernel, but to submit the information to the user mode, and then process the information in the user mode? Thank you.

**Comment by fbs** [comment, marked as answer]
bpftrace uses the perf buffers for all the printing it does, all messages are sent to the userspace part and handled there. We do have `debugf` which uses `printk` but as it ends up in the kernel ringbuf is only if the user requests it

**Reply by caneatcansleep** [reply]
Thank you! I am confused with and interested in this sentence :
> all messages are sent to the userspace part and handled there.

Do you mean that bpftrace automatically extracts data from perf buffer and prints data according to printing format? If so, I have another question:
In Bcc, I can get the data directly from the perf buffer and write codes to process them (in other words, I can get the data and process it according to the characteristics of the data). So the question is how do I get the data from the perf buffer in bpftrace instead of letting bpftrace handle it automatically?

Below is a simple BCC demo. I use `BPF_PERF_OUTPUT(ipv4_events)` to open a perf buffer called `ipv4_events`, and use `b["ipv4_events"].open_perf_buffer(print_event, page_cnt=64)` to get data from perf buffer and the callback function `print_event` will process the data.
```
bpf_text = """
#include <uapi/linux/ptrace.h>
#include <linux/tcp.h>
#include <net/sock.h>
#include <bcc/proto.h>

struct ipv4_data_t {
    u32 pid;
};

BPF_PERF_OUTPUT(ipv4_events);

int kprobe__tcp_v4_rcv(struct pt_regs *ctx)
{
    u32 pid = bpf_get_current_pid_tgid() >> 32;
    struct ipv4_data_t data4 = {};
    data4.pid = pid;
    ipv4_events.perf_submit(ctx, &data4, sizeof(data4));
    return 0;
}
"""

# process event
def print_event(cpu, data, size):
    event = b["ipv4_events"].event(data)
    if event.pid == 30000:
        print("I get the data from perf buffer and I can do whatever I want with the data.")

b = BPF(text=bpf_text)
b["ipv4_events"].open_perf_buffer(print_event, page_cnt=64)
while 1:
    try:
        b.perf_buffer_poll()
    except KeyboardInterrupt:
        exit()
```

**Reply by fbs** [reply]
Internally we currently use perf buffers for all kernel -> userspace communication. So every even that happens on the userspace side is handled by an event sent through it. `printf`, sends a string to print. `print(@map)` sends a 'print this map' event. `exit()` sends a 'terminate' event.

But that's really just an implementation detail. To some extend we don't want the user to worry about this. We want to automatically pick the right solution for their system/kernel. 

To get the data out you have to use `print` or `printf`

**Reply by caneatcansleep** [reply]
Sorry to bother you again. I have been looking at the source code of bpftrace and bcc for the past two days, and I want you to see if the conclusions I got are correct. 
I focused on 1) how bpftrace opens the perf buffer 2) gets data from the buffer and 3) processes the data. Below is the function call chain I got(I'm not quite sure if it is right):
```
main -- bpftrace/src/main.cpp 
    bpftrace.run  -- bpftrace/src/bpftrace.cpp
         setup_perf_events  -- bpftrace/src/bpftrace.cpp
             bpf_open_perf_buffer  -- bcc/src/cc/libbpf.c
                 bpf_open_perf_buffer_opts  -- bcc/src/cc/libbpf.c
         poll_perf_events  -- bpftrace/src/bpftrace.cpp
             perf_reader_event_read  -- bcc/src/cc/perf_reader.c
                 parse_sw 
                     reader->raw_cb=perf_event_printer  -- bpftrace/src/bpftrace.cpp                          
```
Here is my personal opinion：
`perf_event_printer `will be called to process data got from perf buffer.
[perf_event_printer ](https://github.com/iovisor/bpftrace/blob/959970cedef280a408ff9f2a09938a23a6c826b5/src/bpftrace.cpp#L315) implements many functions of bpftrace, such as `print`, `printf`, `clear` and `zero`.

Please check to see whether the above conclusions I got are correct or not. Sincerely look forward to your reply.

**Reply by fbs** [reply]
yes that looks correct but what is your goal here?

**Reply by caneatcansleep** [reply]
According to bpftrace Manual Page, each action block consists of [three parts](https://github.com/iovisor/bpftrace/blob/master/man/adoc/bpftrace.adoc#:~:text=Each%20action%20block%20consists%20of%20three%20parts%3A):
```
probe[,probe]
/predicate/ {
  action
}
```
As you said in your second reply:

> But that's really just an implementation detail. To some extend we don't want the user to worry about this. We want to automatically pick the right solution for their system/kernel.


It seems that bpftrace creates perf buffers, automatically reads data from buffers and automatically processes the data (such as printing).
My thought is can we leave the work of reading data from perf buffer and processing data to the user? That is, bpftrace just output data to perf buffer in `action` part, then the user may write codes( bpftrace, Python with bcc, C with libbpf) to read data from perf buffer and process the data.

**Reply by fbs** [reply]
The issue with what you want is that the messages we pass don't make sense outside of bpftrace itself. Each message has an ID, some have fixed IDs (like exit) but those for printf() are unique for every printf, this avoids sending the format string from the kernel to userspace every time. The rest of the message depends on the ID.
For it to work we would have to develop a plugin API that sends these messages in a structured format to a plugin. I do think there is some value in it but someone will have to add it and we have to work out how to deal with compatibilithy

**Reply by caneatcansleep** [reply]
thank you for your detailed reply, it helps a lot.
